src/Classes/PolygonMask.ts

```diff
@@ -0,0 +1,222 @@
+import {Particle} from "./Particle";
+
+export class PolygonMask {
+    public redrawTimeout?: number;
+    public raw?: number[][];
+    public svg?: SVGSVGElement;
+    public path?: SVGPathElement;
+
+    private readonly container: Container;
+    private width: number;
+    private height: number;
+    private offset?: ICoordinates;
+
+    constructor(container: Container) {
+        this.container = container;
+        this.width = 0;
+        this.height = 0;
+    }
+
+    public checkInsidePolygon(position: ICoordinates | undefined | null) {
+        const container = this.container;
+        const options = container.options;
+
+        // https://github.com/substack/point-in-polygon
+        // ray-casting algorithm based on
+        // http://www.ecse.rpi.edu/Homepages/wrf/Research/Short_Notes/pnpoly.html
+        if (options.polygon.type !== PolygonMaskType.inline) {
+            if (!this.raw) {
+                this.raw = [];
+            }
+
+            if (this.raw.length > 0) {
+                const x = position ? position.x : Math.random() * container.canvas.width;
+                const y = position ? position.y : Math.random() * container.canvas.height;
+                let inside = false;
+
+                for (let i = 0, j = this.raw.length - 1; i < this.raw.length; j = i++) {
+                    const xi = this.raw[i][0], yi = this.raw[i][1];
+                    const xj = this.raw[j][0], yj = this.raw[j][1];
+                    const intersect = ((yi > y) != (yj > y)) && (x < (xj - xi) * (y - yi) / (yj - yi) + xi);
+
+                    if (intersect) {
+                        inside = !inside;
+                    }
+                }
+
+                if (options.polygon.type === PolygonMaskType.inside) {
+                    return inside;
+                } else if (options.polygon.type === PolygonMaskType.outside) {
+                    return !inside;
+                }
+            } else {
+                console.error('No polygon found, you need to specify SVG url in config.');
+                return true;
+            }
+        } else {
+            return true;
+        }
+    }
+
+    public randomPointInPolygon(): ICoordinates {
+        const container = this.container;
+        const position = {
+            x: Math.random() * container.canvas.width,
+            y: Math.random() * container.canvas.height,
+        };
+
+        if (this.checkInsidePolygon(position)) {
+            return position;
+        } else {
+            return this.randomPointInPolygon();
+        }
+    }
+
+    /**
+     * Depends on SVGPathSeg API polyfill https://github.com/progers/pathseg for Chrome
+     * Deprecate SVGPathElement.getPathSegAtLength removed in:
+     * Chrome for desktop release 62
+     * Chrome for Android release 62
+     * Android WebView release 62
+     * Opera release 49
+     * Opera for Android release 49
+     */
+    public async parseSvgPathToPolygon(svgUrl?: string) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [expected call-signature: 'parseSvgPathToPolygon' to have a typedef](https://app.codacy.com/manual/ar3s/tsparticles/pullRequest?prid=5042590)